rust/src/mat.rs

```diff
@@ -1,93 +1,196 @@
+
+        let row_major_stride: Vec<usize> = vec![
+            z_length * y_length * self.voxel_size,
+            z_length * self.voxel_size,
+            self.voxel_size,
+        ];
+        let column_major_stride: Vec<usize> = vec![
+            self.voxel_size,
+            x_length * self.voxel_size,
+            x_length * y_length * self.voxel_size,
+        ];
+        // Do continuous read in z. Last dim in Row-Major is continuous.
+        for x in 0usize..x_length {
+            for y in 0usize..y_length {
+                for z in 0usize..z_length {
+                    let row_major_index =
+                        x * row_major_stride[0] + y * row_major_stride[1] + z * row_major_stride[2];
+                    let column_major_index = x * column_major_stride[0]
+                        + y * column_major_stride[1]
+                        + z * column_major_stride[2];
+                    for byte_offset in 0..self.voxel_size {
+                        buffer_data[column_major_index as usize + byte_offset as usize] =
+                            self.data[row_major_index as usize + byte_offset as usize];
+                    }
+                }
+            }
+        }
+        Ok(())
     }
 
     pub fn copy_from(&mut self, dst_pos: Vec3, src: &Mat, src_box: Box3) -> Result<()> {
         // make sure that matrices are matching
-        if self.voxel_size != src.voxel_size { return Err("Matrices mismatch in voxel size"); }
-        if self.voxel_type != src.voxel_type { return Err("Matrices mismatch in voxel type"); }
-
-        if !(src_box.max() < (src.shape + 1)) { return Err("Reading out of bounds"); }
-        if !(dst_pos + src_box.width() < (self.shape + 1)) { return Err("Writing out of bounds"); }
-
-        let len = src_box.width();
-        let stripe_len = src.voxel_size * len.x as usize;
-
-        let src_off_y = (src.shape.x as usize * src.voxel_size) as isize;
-        let src_off_z = (src.shape.x as usize * src.shape.y as usize * self.voxel_size) as isize;
+        if self.voxel_size != src.voxel_size {
+            return Err("Matrices mismatch in voxel size");
+        }
+        if self.voxel_type != src.voxel_type {
+            return Err("Matrices mismatch in voxel type");
+        }
+        if !(src_box.max() < (src.shape + 1)) {
+            return Err("Reading out of bounds");
+        }
+        if !(dst_pos + src_box.width() < (self.shape + 1)) {
+            return Err("Writing out of bounds");
+        }
+        if self.is_fortran_order != src.get_is_fortran_order() {
+            return Err("source and destination has to be the same order");
+        }
 
-        let dst_off_y = (self.shape.x as usize * self.voxel_size) as isize;
-        let dst_off_z = (self.shape.x as usize * self.shape.y as usize * self.voxel_size) as isize;
+        let width = src_box.width();
+        // unified has fast to slow moving indices
+        let unified_width = match self.is_fortran_order {
```

Comment:
It seems quite weird to me to use `match` on a boolean, when you could also write `let unified_width = if self.is_fortran_order { ... } else { ... }` (shaves off the `true`/`false` literals). But I don't know jack about idiomatic rust, so I might be wrong here :smile: Same in the following blocks..